odoo/domains.py

```diff
@@ -0,0 +1,1467 @@
+        try:
+            children = list(self._simplify_children(c._optimize(model) for c in children))
+            while True:
+                if len(children) < 2:
+                    break
+                optimal_children.update(id(c) for c in children)
+                for opt in _NARY_OPTIMIZATIONS:
+                    children = opt(cls, children, model)
+                optimized = True
+                for i, child in enumerate(children):
+                    if id(child) in optimal_children:
+                        continue
+                    children[i] = child._optimize(model)
+                    optimized = False
+                if optimized:
+                    break
+                children = list(self._simplify_children(children))
+        except DomainConst as c:
+            return c
+        # return the result a as simple node
+        len_children = len(children)
+        if len_children == 0:
+            return self.ZERO
+        if len_children == 1:
+            return children[0]
+        return cls(children, _model_name=model._name)
+
+    def _build_sql(self, model: BaseModel, alias: str, query: Query) -> SQL:
+        assert self.children, "No children, optimize() probably not executed"
+        return SQL("(%s)", self.OPERATOR_SQL.join(
+            c._build_sql(model, alias, query) for c in self.children
+        ))
+
+    @classmethod
+    def build_sql(cls, *items: SQL) -> SQL:
+        """Build the SQL value combining the given items"""
+        if not items:
+            return SQL(str(cls.ZERO.VALUE).upper())
+        if len(items) == 1:
+            return items[0]
+        return SQL("(%s)", cls.OPERATOR_SQL.join(items))
+
+
+class DomainAnd(DomainNary):
+    """Domain: AND with multiple children"""
+    OPERATOR = '&'
+    OPERATOR_SQL = SQL(" AND ")
+    ZERO = TRUE_DOMAIN
+
+    def __invert__(self):
+        return DomainOr([~c for c in self.children])
+
+    def __and__(self, other):
+        # simple optimization to append children
+        if isinstance(other, DomainAnd):
+            return DomainAnd(self.children + other.children)
+        if isinstance(other, DomainBase) and not isinstance(other, (DomainConst, DomainNary)):
+            return DomainAnd(self.children + [other])
+        return super().__and__(other)
+
+
+class DomainOr(DomainNary):
+    """Domain: OR with multiple children"""
+    OPERATOR = '|'
+    OPERATOR_SQL = SQL(" OR ")
+    ZERO = FALSE_DOMAIN
+
+    def __invert__(self):
+        return DomainAnd([~c for c in self.children])
+
+    def __or__(self, other):
+        # simple optimization to append children
+        if isinstance(other, DomainOr):
+            return DomainOr(self.children + other.children)
+        if isinstance(other, DomainBase) and not isinstance(other, (DomainConst, DomainNary)):
+            return DomainOr(self.children + [other])
+        return super().__or__(other)
+
+
+class DomainField(DomainBase):
```

Comment:
I could use `DomainLeaf`. It seemed a little bit generic, because maybe in the future we would like to distinguish (field, operator, value) from (expression, operator, value) - and refactor or subclass in the future if needed.
For naming simplicity, I started all classes with `Domain...`.